### PR TITLE
[Go] Generate inline enums

### LINKED ISSUE #2360: [BUG][Go][Python] Inline enums declared as simple strings
#### Bug Report Checklist

- [ x ] Have you provided a full/minimal spec to reproduce the issue?
- [ x ] Have you validated the input using an OpenAPI validator ([example](https://apidevtools.org/swagger-parser/online/))?
- [ x ] What's the version of OpenAPI Generator used?
- [ x ] Have you search for related issues/PRs?
- [ x ] What's the actual output vs expected output?

##### Description

Enums that are declared inline are simply being declared as strings, while the same enum defined as a separate schema is properly captured with all legal values. This appears to be across many generators, I see it in at least `go` and `python`.

##### openapi-generator version

`4.0.0-SNAPSHOT` built from this commit: https://github.com/OpenAPITools/openapi-generator/commit/b128d1470709c44c6c7b6a906e1993d1c2758b52

##### OpenAPI declaration file content or url

sample.yaml:
```yaml
openapi: "3.0.0"
info:
  version: v1
  title: Fake Service
servers:
- url: https://fake.net
paths:
  /status/details:
    get:
      summary: Get status details
      operationId: getStatus
      responses:
        200:
          description: Success
          content:
            application/json:
              schema:
                type: array
                items:
                  $ref: "#/components/schemas/StatusDetails"

components:
  schemas:
    State:
      type: string
      enum:
      - "IN_PROGRESS"
      - "READY"
      - "FAILED"
    StatusDetails:
      properties:
        id:
          type: string
        stateInline: # enum not captured, just declared as "string"
          type: string
          enum:
          - "IN_PROGRESS"
          - "READY"
          - "FAILED"
        state: # works as expected
          $ref: "#/components/schemas/State"
```

##### Command line used for generation

`java -jar openapi-generator-cli.jar generate -i sample.yaml -g go -o ./sample-go`
`java -jar openapi-generator-cli.jar generate -i sample.yaml -g python -o ./sample-python`

##### Steps to reproduce

1. `git clone https://github.com/OpenAPITools/openapi-generator.git`
2. `cd openapi-generator`
3. `git checkout b128d1470709c44c6c7b6a906e1993d1c2758b52`
4. `mvn clean install`
5. `ln -s modules/openapi-generator-cli/target/openapi-generator-cli.jar .`
6. `java -jar openapi-generator-cli.jar generate -i sample.yaml -g go -o ./sample-go` or `java -jar openapi-generator-cli.jar generate -i sample.yaml -g python -o ./sample-python`

##### Related issues/PRs

https://github.com/OpenAPITools/openapi-generator/issues/2200
https://github.com/OpenAPITools/openapi-generator/issues/1702
https://github.com/OpenAPITools/openapi-generator/issues/1706

##### Actual vs Expected output

Go:
model_status_details.go (actual):
```go
package openapi

type StatusDetails struct {
	Id string `json:"id,omitempty"`
	StateInline string `json:"stateInline,omitempty"`
	State State `json:"state,omitempty"`
}
```

model_status_details.go (expected):
```go
package openapi

type StatusDetails struct {
	Id string `json:"id,omitempty"`
	StateInline StateInlineEnum `json:"stateInline,omitempty"`
	State State `json:"state,omitempty"`
}

type StateInlineEnum string

// List of StateInlineEnum
const (
	IN_PROGRESS StateInlineEnum = "IN_PROGRESS"
	READY StateInlineEnum = "READY"
	FAILED StateInlineEnum = "FAILED"
)
```

Python:
openapiclient/models/status_details.py (actual, abbreviated):
```python
...
class StatusDetails(object):
    """NOTE: This class is auto generated by OpenAPI Generator.
    Ref: https://openapi-generator.tech

    Do not edit the class manually.
    """

    """
    Attributes:
      openapi_types (dict): The key is attribute name
                            and the value is attribute type.
      attribute_map (dict): The key is attribute name
                            and the value is json key in definition.
    """
    openapi_types = {
        'id': 'str',
        'state_inline': 'str',
        'state': 'State'
    }
```

openapiclient/models/status_details.py (expected, abbreviated):
```python
...
class StatusDetails(object):
    """NOTE: This class is auto generated by OpenAPI Generator.
    Ref: https://openapi-generator.tech

    Do not edit the class manually.
    """

    """
    Attributes:
      openapi_types (dict): The key is attribute name
                            and the value is attribute type.
      attribute_map (dict): The key is attribute name
                            and the value is json key in definition.
    """
    openapi_types = {
        'id': 'str',
        'state_inline': 'StateInlineEnum',
        'state': 'State'
    }
```

##### Suggest a fix

If inline enum schema models should be formed the same as those defined as separate components.

### PR DESCRIPTION
**Pull request by ctreatma**
This adds a new `model_inner_enum.mustache` template to the Go client generator and updates the `model_simple.mustache` template to render `model_inner_enum` inline for any variable that is an enum.

This is a template-only change that I'm using in a different project; the template change is pretty hairy and it feels like a fair amount of it should be moved into the generator code itself (maybe `GoClientCodegen.java`, but maybe it's useful to have it in `AbstractGoCodegen.java` for other Go generators?).  I'm not sure how best to do that, though; maybe update the generator to create separate `model` objects for inline enums and then render them the same way we currently render external enums?

Fixes #2360

### PR checklist
 
- [x] Read the [contribution guidelines](https://github.com/openapitools/openapi-generator/blob/master/CONTRIBUTING.md).
- [x] Pull Request title clearly describes the work in the pull request and Pull Request description provides details about how to validate the work. Missing information here may result in delayed response from the community.
- [x] Run the following to [build the project](https://github.com/OpenAPITools/openapi-generator#14---build-projects) and update samples:
  ```
  ./mvnw clean package 
  ./bin/generate-samples.sh
  ./bin/utils/export_docs_generators.sh
  ``` 
  Commit all changed files. 
  This is important, as CI jobs will verify _all_ generator outputs of your HEAD commit as it would merge with master. 
  These must match the expectations made by your contribution. 
  You may regenerate an individual generator by passing the relevant config(s) as an argument to the script, for example `./bin/generate-samples.sh bin/configs/java*`. 
  For Windows users, please run the script in [Git BASH](https://gitforwindows.org/).
- [ ] In case you are adding a new generator, run the following additional script : 
  ```
  ./bin/utils/ensure-up-to-date
  ``` 
  Commit all changed files.
- [x] File the PR against the [correct branch](https://github.com/OpenAPITools/openapi-generator/wiki/Git-Branches): `master` (7.0.1 - patch release), `7.1.x` (minor release - breaking changes with fallbacks), `8.0.x` (breaking changes without fallbacks)
- [ ] If your PR is targeting a particular programming language, @mention the [technical committee](https://github.com/openapitools/openapi-generator/#62---openapi-generator-technical-committee) members, so they are more likely to review the pull request.

### CODE REVIEW

File: modules/openapi-generator/src/main/resources/go/model_inner_enum.mustache [hunk 1]
```diff
@@ -0,0 +1,101 @@
+// {{{classname}}}{{#lambda.pascalcase}}{{{nameInCamelCase}}}{{/lambda.pascalcase}} {{{description}}}{{^description}}the model '{{{classname}}}{{#lambda.pascalcase}}{{{nameInCamelCase}}}{{/lambda.pascalcase}}'{{/description}}
+type {{{classname}}}{{#lambda.pascalcase}}{{{nameInCamelCase}}}{{/lambda.pascalcase}} {{{format}}}{{^format}}{{dataType}}{{/format}}
```
Comment (ctreatma) [inline modules/openapi-generator/src/main/resources/go/model_inner_enum.mustache:2]: A lot of this seems like stuff we might get for free if I could figure out how/where in `GoClientCodegen.java` to recognize that a `param.isEnum` and then create a model for that enum and update the param to reference the enum.

File: modules/openapi-generator/src/main/resources/go/model_simple.mustache [hunk 2]
```diff
@@ -196,13 +209,13 @@ func (o *{{classname}}) Get{{name}}() {{vendorExtensions.x-go-base-type}} {
...
-func (o *{{classname}}) Get{{name}}Ok() ({{^isArray}}{{^isFreeFormObject}}*{{/isFreeFormObject}}{{/isArray}}{{vendorExtensions.x-go-base-type}}, bool) {
+func (o *{{classname}}) Get{{name}}Ok() ({{^isArray}}{{^isFreeFormObject}}*{{/isFreeFormObject}}{{/isArray}}{{#isEnum}}{{{classname}}}{{#lambda.pascalcase}}{{{nameInCamelCase}}}{{/lambda.pascalcase}}{{/isEnum}}{{^isEnum}}{{{vendorExtensions.x-go-base-type}}}{{/isEnum}}, bool) {
 	if o == nil{{^isNullable}} || IsNil(o.{{name}}){{/isNullable}}{{#isNullable}}{{#vendorExtensions.x-golang-is-container}} || IsNil(o.{{name}}){{/vendorExtensions.x-golang-is-container}}{{/isNullable}} {
 	{{^isFreeFormObject}}
-		return nil, false
+		return {{#isEnum}}o.{{name}}{{/isEnum}}{{^isEnum}}nil{{/isEnum}}, false
 	{{/isFreeFormObject}}
 	{{#isFreeFormObject}}
-		return {{vendorExtensions.x-go-base-type}}{}, false
+		return {{#isEnum}}{{{classname}}}{{#lambda.pascalcase}}{{{nameInCamelCase}}}{{/lambda.pascalcase}}{{/isEnum}}{{^isEnum}}{{{vendorExtensions.x-go-base-type}}}{{/isEnum}}{}, false
```
Comment (ctreatma) [inline modules/openapi-generator/src/main/resources/go/model_simple.mustache:218]: I probably changed the template in places where it's not necessary (for example, I don't know if `#isFreeFormObject` implies that the var is not an inline enum); it works for my purposes but would need to be cleaned up _if_ we can't instead update the generator to create separate models for inline enums

Comment (wing328) [reply modules/openapi-generator/src/main/resources/go/model_simple.mustache:218]: To create inline enum as separate models, there's an option in inline schema resolver to do that 

https://github.com/OpenAPITools/openapi-generator/blob/master/docs/customization.md#inline-schema-naming => RESOLVE_INLINE_ENUMS=true

File: modules/openapi-generator/src/main/resources/go/model_inner_enum.mustache (hunk 1 above)
Comment (wing328) [reply modules/openapi-generator/src/main/resources/go/model_inner_enum.mustache:2]: please see below: https://github.com/OpenAPITools/openapi-generator/pull/16582#discussion_r1325278796

File: modules/openapi-generator/src/main/resources/go/model_simple.mustache (hunk 2 above)
Comment (ctreatma) [reply modules/openapi-generator/src/main/resources/go/model_simple.mustache:218]: Oh dang!  That's way better.  Thanks!

**Comment by ctreatma** [comment]
Closing this PR since the inline enum issue is already fixed in `v7.0.0` with `RESOLVE_INLINE_ENUMS=true`

**Pull request closed without merging**